File: backend/catalog-service/app/services/show_service.py
from typing import List, Optional
from datetime import date

from app.models import Show
from app.repositories import ShowRepository, MovieRepository, SeatRepository
from app.schemas import CreateShowRequest


class ShowService:
    def __init__(
        self,
        repository: ShowRepository,
        movie_repository: MovieRepository,
        seat_repository: SeatRepository,
    ):
        self.repository = repository
        self.movie_repository = movie_repository
        self.seat_repository = seat_repository

    async def create_show(self, request: CreateShowRequest) -> Show:
        # Verify movie exists
        movie = await self.movie_repository.get_by_id(request.movie_id)
        if not movie:
            raise ValueError(f"Movie with id {request.movie_id} not found")
        
        show = Show(
            movie_id=request.movie_id,
            screen_id=request.screen_id,
            start_time=request.start_time,
            end_time=request.end_time,
            base_price=request.base_price,
        )
        
        return await self.repository.create(show)

    async def get_show_by_id(self, show_id: int) -> Optional[Show]:
        return await self.repository.get_by_id(show_id)

    async def get_shows_by_movie(self, movie_id: int) -> List[Show]:
        return await self.repository.get_by_movie_id(movie_id)

    async def get_shows_by_movie_and_city(self, movie_id: int, city: str) -> List[Show]:
        return await self.repository.get_by_movie_and_city(movie_id, city)

    async def get_shows_by_date(self, show_date: date) -> List[Show]:
        return await self.repository.get_by_date(show_date)

    async def get_shows_by_movie_city_and_date(
        self, movie_id: int, city: str, show_date: date
    ) -> List[Show]:
        return await self.repository.get_by_movie_city_and_date(movie_id, city, show_date)

    async def delete_show(self, show_id: int) -> None:
        show = await self.repository.get_by_id(show_id)
        if not show:
            raise ValueError(f"Show with id {show_id} not found")
        
        await self.repository.delete(show)

File: backend/catalog-service/src/main/java/com/bookmyshow/catalog/service/ShowService.java
package com.bookmyshow.catalog.service;

import com.bookmyshow.catalog.dto.CreateShowRequest;
import com.bookmyshow.catalog.entity.Movie;
import com.bookmyshow.catalog.entity.Screen;
import com.bookmyshow.catalog.entity.Show;
import com.bookmyshow.catalog.repository.MovieRepository;
import com.bookmyshow.catalog.repository.ScreenRepository;
import com.bookmyshow.catalog.repository.ShowRepository;
import com.bookmyshow.common.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

@Service
@Transactional
@RequiredArgsConstructor
public class ShowService {

    private final ShowRepository showRepository;
    private final MovieRepository movieRepository;
    private final ScreenRepository screenRepository;

    public Show createShow(CreateShowRequest request) {
        Movie movie = movieRepository.findById(request.getMovieId())
                .orElseThrow(() -> new ResourceNotFoundException("Movie", "id", request.getMovieId()));

        Screen screen = screenRepository.findById(request.getScreenId())
                .orElseThrow(() -> new ResourceNotFoundException("Screen", "id", request.getScreenId()));

        Show show = new Show();
        show.setMovie(movie);
        show.setScreen(screen);
        show.setStartTime(request.getStartTime());
        show.setEndTime(request.getEndTime());
        show.setBasePrice(request.getBasePrice());

        return showRepository.save(show);
    }

    public Show getShowById(Long id) {
        return showRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Show", "id", id));
    }

    public List<Show> getShowsByMovie(Long movieId) {
        return showRepository.findByMovieId(movieId);
    }

    public List<Show> getShowsByMovieAndCity(Long movieId, String city) {
        return showRepository.findByMovieIdAndCity(movieId, city);
    }

    public List<Show> getShowsByDate(LocalDate date) {
        return showRepository.findByDate(date);
    }

    public List<Show> getShowsByMovieAndCityAndDate(Long movieId, String city, LocalDate date) {
        return showRepository.findByMovieIdAndCityAndDate(movieId, city, date);
    }

    public void deleteShow(Long id) {
        Show show = getShowById(id);
        showRepository.delete(show);
    }
}
